### DISCUSSION TITLE
migrating from CentOS7 to Rocky9 /  rank mismatch / more actual than formal arguments

### DISCUSSION THREAD

**Discussion by gsal** [CGNS/CGNS#735] [CGNStalk]
So, we lost a developer who wrote a program using CGNS 3.3.1, while in CentOS 7; now, we are moving to Rocky 9, I want to re-compile (and deploy) in Rocky9 both CGNS and the program. 

Attempt 1: 
Compile CGNS 3.3.1 in Rocky 9...unsuccesful; I get a couple of the following errors: 

```
gfortran -Wno-c-binding-type  -I.  -DBUILD_HDF5 -o lib/cgns_f.o -c cgns_f.F90
cgns_f.F90:4964:54:

 4964 |     CALL cgio_get_dimensions_f_c(cgio_num, id, ndims, dims, ier)
      |                                                      1
......
 4976 |     CALL cgio_get_dimensions_f_c(cgio_num, id, ndims, dims, ier)
      |                                                      2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)
```

Attempt 2:
Compile CGNS 3.4.2 in Rocky 9 ...succesful
Compile program...unsuccesful, I get about a dozen of these error messages: 
```
m2_cgns.f90:280:71:

  280 |   CALL cg_goto_f(file_index,base_index,ier,'FlowEquationSet_t',1,'end')
      |                                                                       1
Error: More actual than formal arguments in procedure call at (1)
m2_cgns.f90:284:86:

  284 |   CALL cg_goto_f(file_index,base_index,ier,'FlowEquationSet_t',1,'GasModel_t',1,'end')
      |                                                                                      1
Error: More actual than formal arguments in procedure call at (1)
m2_cgns.f90:291:71:

  291 |   CALL cg_goto_f(file_index,base_index,ier,'FlowEquationSet_t',1,'end')
      |                                                                       1
Error: More actual than formal arguments in procedure call at (1)
```

Any hints?

**Comment by brtnfld** [comment]
This has been fixed in the new release of CGNS, and you need to add the Fortran compiler flag -fallow-argument-mismatch

On Oct 1, 2023, at 7:37 PM, gsal ***@***.***> wrote:



So, we lost a developer who wrote a program using CGNS 3.3.1, while in CentOS 7; now, we are moving to Rocky 9, I want to re-compile (and deploy) in Rocky9 both CGNS and the program.

Attempt 1:
Compile CGNS 3.3.1 in Rocky 9...unsuccesful; I get a couple of the following errors:

gfortran -Wno-c-binding-type  -I.  -DBUILD_HDF5 -o lib/cgns_f.o -c cgns_f.F90
cgns_f.F90:4964:54:

 4964 |     CALL cgio_get_dimensions_f_c(cgio_num, id, ndims, dims, ier)
      |                                                      1
......
 4976 |     CALL cgio_get_dimensions_f_c(cgio_num, id, ndims, dims, ier)
      |                                                      2
Error: Rank mismatch between actual argument at (1) and actual argument at (2) (rank-1 and scalar)


Attempt 2:
Compile CGNS 3.4.2 in Rocky 9 ...succesful
Compile program...unsuccesful, I get about a dozen of these error messages:

m2_cgns.f90:280:71:

  280 |   CALL cg_goto_f(file_index,base_index,ier,'FlowEquationSet_t',1,'end')
      |                                                                       1
Error: More actual than formal arguments in procedure call at (1)
m2_cgns.f90:284:86:

  284 |   CALL cg_goto_f(file_index,base_index,ier,'FlowEquationSet_t',1,'GasModel_t',1,'end')
      |                                                                                      1
Error: More actual than formal arguments in procedure call at (1)
m2_cgns.f90:291:71:

  291 |   CALL cg_goto_f(file_index,base_index,ier,'FlowEquationSet_t',1,'end')
      |                                                                       1
Error: More actual than formal arguments in procedure call at (1)


Any hints?

**Reply by gsal** [reply]
fixed in the new release?  As in 4.3.0? 
I fear that if I jump from using 3.3.1 to 4.3.0,  there will for sure be some compatibility issues that I may not become aware of and/or if I do, won't know how to fix.....that's why I was trying to stay within versions 3.y.z

Anyway, will try the fortran flag and see how far I get. I can also try the latest release.

**Reply by gsal** [reply]
When attempting to buil CGNS 3.3.1 in rocky9, that fortran flag helped get pass the rank mismatch problem; but, then, I run into a couple of other problems, one being `mattherr` that seems went obsolete in "recent" GCC versions...and rocky9 comes with GCC 11.
 
Any ideas on how to get passed the 
```
Error: More actual than formal arguments in procedure call at (1)
```
I am getting this when attempting to compile the one program, even when linking against CGNS 4.4.0

I can see in cgns_f.F90 that cg_goto_f() is suposed to handle up to 20 UserDataNames; but, somehow, when my program call the subroutine, it can't take even 1 !

**Comment by brtnfld** [comment]
There should not be an issue with upgrading to 4.3.0, which has numerous benefits over 3.3.1. I compile CGNS with gcc-12 with no problems. I'm not sure what `mattherr` is.  

What line is it complaining about? Do you know if the CGNS test all passed on your installed version? Do you have a simple reproducer?

**Comment by gsal** [comment]
Typo...it is not `mattherr`, it's just `matherr`.  I run into the issued referred to [here](https://bugzilla.redhat.com/show_bug.cgi?id=1541616) and [here](https://bugs.gentoo.org/662210).

Anyway, I fixed the problem in the program when linking CGNS 4.4.0.  

I described the problem with the in-house program, above...every time `cg_goto_f`  was called, I would get a compile error saying "too many arguments".  In comparing the sources between CGNS 3.3.1 and 4.4.0, I notice a difference in some explicit interface for `cg_goto_f`, so, I ventured into the in-house program sources and added `USE CGNS  cg_goto_f `....this solved my problem. 

Thanks for your assistance, I helped to hear that I can safely jumped to CGNS 4.y
